File: extension/src/score/listingDrift.ts
import { cosineSimilarity, EMBEDDING_DIMENSIONS, embedText } from "./textEmbedding";


export interface ReviewForDrift {
  text: string | null;
}

export interface ListingDriftOptions {
  dimensions?: number;
  offTopicDistance?: number;
  embeddingCache?: WeakMap<ReviewForDrift, number[]>;
}

export interface ChangePoint {
  day: number;
  afterCount: number;
}

export interface ListingDriftResult {
  offTopicShare: number | null;
  offTopicCount: number;
  meanDistance: number | null;
  changePoint: ChangePoint | null;
  driftStatistic: number;
  embeddedCount: number;
}

export const DEFAULT_OFF_TOPIC_DISTANCE = 1;

export const MINIMUM_CHANGE_POINT_REVIEWS = 12;
export const MINIMUM_SEGMENT_REVIEWS = 5;

export const MINIMUM_REPORTABLE_DRIFT = 0.5;

const ABSENT: ListingDriftResult = {
  offTopicShare: null,
  offTopicCount: 0,
  meanDistance: null,
  changePoint: null,
  driftStatistic: 0,
  embeddedCount: 0,
};

function embeddingFor(
  review: ReviewForDrift,
  dimensions: number,
  cache: WeakMap<ReviewForDrift, number[]> | undefined,
): number[] | null {
  const cached = cache?.get(review);
  if (cached !== undefined && cached.length === dimensions) {
    return cached;
  }
  if (review.text === null || review.text.length === 0) {
    return null;
  }
  const embedding = embedText(review.text, dimensions);
  if (embedding !== null) {
    cache?.set(review, embedding);
  }
  return embedding;
}

export function listingIdentityDrift(
  reviews: readonly ReviewForDrift[],
  days: readonly (number | null)[],
  productText: string,
  options: ListingDriftOptions = {},
): ListingDriftResult {
  const dimensions = options.dimensions ?? EMBEDDING_DIMENSIONS;
  const offTopicDistance = options.offTopicDistance ?? DEFAULT_OFF_TOPIC_DISTANCE;

  const embedded: { embedding: number[]; day: number | null }[] = [];
  for (const [index, review] of reviews.entries()) {
    const embedding = embeddingFor(review, dimensions, options.embeddingCache);
    if (embedding !== null) {
      embedded.push({ embedding, day: days[index] ?? null });
    }
  }
  if (embedded.length === 0) {
    return ABSENT;
  }

  const product = embedText(productText, dimensions);
  let offTopicCount = 0;
  let meanDistance: number | null = null;
  if (product !== null) {
    let total = 0;
    for (const entry of embedded) {
      const distance = 1 - cosineSimilarity(entry.embedding, product);
      total += distance;
      if (distance >= offTopicDistance) {
        offTopicCount++;
      }
    }
    meanDistance = total / embedded.length;
  }

  const dated = embedded
    .filter((entry): entry is { embedding: number[]; day: number } => entry.day !== null)
    .sort((a, b) => a.day - b.day);
  const { changePoint, driftStatistic } = centroidChangePoint(dated);

  return {
    offTopicShare: product === null ? null : offTopicCount / embedded.length,
    offTopicCount,
    meanDistance,
    changePoint,
    driftStatistic,
    embeddedCount: embedded.length,
  };
}

export function centroidChangePoint(
  dated: readonly { embedding: number[]; day: number }[],
): { changePoint: ChangePoint | null; driftStatistic: number } {
  const count = dated.length;
  if (count < MINIMUM_CHANGE_POINT_REVIEWS) {
    return { changePoint: null, driftStatistic: 0 };
  }
  const dimensions = (dated[0] as { embedding: number[] }).embedding.length;

  const total = new Array<number>(dimensions).fill(0);
  for (const entry of dated) {
    for (let i = 0; i < dimensions; i++) {
      total[i] = (total[i] as number) + (entry.embedding[i] as number);
    }
  }

  const before = new Array<number>(dimensions).fill(0);
  let widest = 0;
  let splitIndex = -1;
  for (let k = 1; k < count; k++) {
    const embedding = (dated[k - 1] as { embedding: number[] }).embedding;
    for (let i = 0; i < dimensions; i++) {
      before[i] = (before[i] as number) + (embedding[i] as number);
    }
    if (k < MINIMUM_SEGMENT_REVIEWS || k > count - MINIMUM_SEGMENT_REVIEWS) {
      continue;
    }
    let dot = 0;
    let beforeSquares = 0;
    let afterSquares = 0;
    for (let i = 0; i < dimensions; i++) {
      const beforeValue = before[i] as number;
      const afterValue = (total[i] as number) - beforeValue;
      dot += beforeValue * afterValue;
      beforeSquares += beforeValue * beforeValue;
      afterSquares += afterValue * afterValue;
    }
    if (beforeSquares === 0 || afterSquares === 0) {
      continue;
    }
    const distance = 1 - dot / Math.sqrt(beforeSquares * afterSquares);
    if (distance > widest) {
      widest = distance;
      splitIndex = k;
    }
  }

  if (splitIndex < 0 || widest < MINIMUM_REPORTABLE_DRIFT) {
    return { changePoint: null, driftStatistic: widest };
  }
  return {
    changePoint: {
      day: (dated[splitIndex] as { day: number }).day,
      afterCount: count - splitIndex,
    },
    driftStatistic: widest,
  };
}

File: research/verdict_research/features/listing_drift.py
import math
from dataclasses import dataclass

from verdict_research.features.text_embedding import (
    EMBEDDING_DIMENSIONS,
    cosine_similarity,
    embed_text,
)

DEFAULT_OFF_TOPIC_DISTANCE = 1.0

MINIMUM_CHANGE_POINT_REVIEWS = 12
MINIMUM_SEGMENT_REVIEWS = 5

MINIMUM_REPORTABLE_DRIFT = 0.5


@dataclass
class ReviewForDrift:
    text: str | None


@dataclass
class ChangePoint:
    day: int
    after_count: int


@dataclass
class ListingDriftResult:
    off_topic_share: float | None
    off_topic_count: int
    mean_distance: float | None
    change_point: ChangePoint | None
    drift_statistic: float
    embedded_count: int


_ABSENT = ListingDriftResult(
    off_topic_share=None,
    off_topic_count=0,
    mean_distance=None,
    change_point=None,
    drift_statistic=0.0,
    embedded_count=0,
)


def listing_identity_drift(
    reviews: list[ReviewForDrift],
    days: list[int | None],
    product_text: str,
    dimensions: int = EMBEDDING_DIMENSIONS,
    off_topic_distance: float = DEFAULT_OFF_TOPIC_DISTANCE,
) -> ListingDriftResult:
    embedded: list[tuple[list[float], int | None]] = []
    for index, review in enumerate(reviews):
        if review.text is None or review.text == "":
            continue
        embedding = embed_text(review.text, dimensions)
        if embedding is not None:
            embedded.append((embedding, days[index]))
    if not embedded:
        return _ABSENT

    product = embed_text(product_text, dimensions)
    off_topic_count = 0
    mean_distance = None
    if product is not None:
        total = 0.0
        for embedding, _ in embedded:
            distance = 1 - cosine_similarity(embedding, product)
            total += distance
            if distance >= off_topic_distance:
                off_topic_count += 1
        mean_distance = total / len(embedded)

    dated = sorted(
        ((embedding, day) for embedding, day in embedded if day is not None),
        key=lambda entry: entry[1],
    )
    change_point, drift_statistic = centroid_change_point(dated)

    return ListingDriftResult(
        off_topic_share=None if product is None else off_topic_count / len(embedded),
        off_topic_count=off_topic_count,
        mean_distance=mean_distance,
        change_point=change_point,
        drift_statistic=drift_statistic,
        embedded_count=len(embedded),
    )


def centroid_change_point(
    dated: list[tuple[list[float], int]],
) -> tuple[ChangePoint | None, float]:
    count = len(dated)
    if count < MINIMUM_CHANGE_POINT_REVIEWS:
        return None, 0.0
    dimensions = len(dated[0][0])

    total = [0.0] * dimensions
    for embedding, _ in dated:
        for i in range(dimensions):
            total[i] += embedding[i]

    before = [0.0] * dimensions
    widest = 0.0
    split_index = -1
    for k in range(1, count):
        embedding = dated[k - 1][0]
        for i in range(dimensions):
            before[i] += embedding[i]
        if k < MINIMUM_SEGMENT_REVIEWS or k > count - MINIMUM_SEGMENT_REVIEWS:
            continue
        dot = 0.0
        before_squares = 0.0
        after_squares = 0.0
        for i in range(dimensions):
            before_value = before[i]
            after_value = total[i] - before_value
            dot += before_value * after_value
            before_squares += before_value * before_value
            after_squares += after_value * after_value
        if before_squares == 0 or after_squares == 0:
            continue
        distance = 1 - dot / math.sqrt(before_squares * after_squares)
        if distance > widest:
            widest = distance
            split_index = k

    if split_index < 0 or widest < MINIMUM_REPORTABLE_DRIFT:
        return None, widest
    return ChangePoint(day=dated[split_index][1], after_count=count - split_index), widest
